packages/synapse-react-client/src/components/DataGrid/DataGrid.tsx

```diff
@@ -0,0 +1,695 @@
+      performCommit(newValue)
+    }, 500),
+  )
+
+  function addRowToModel() {
+    return { _rowId: genId() }
+  }
+
+  return (
+    <div>
+      <div>
+        <input
+          ref={inputRef}
+          type="text"
+          placeholder="(Optional) Enter a SQL query or session ID"
+          style={{ width: '300px' }}
+          onKeyDown={e => {
+            if (e.key === 'Enter') {
+              const inputValue = inputRef.current?.value || ''
+              handleStartSession(inputValue)
+            }
+          }}
+        />
+        <button
+          onClick={() => {
+            const inputValue = inputRef.current?.value || ''
+            handleStartSession(inputValue)
+          }}
+          style={{ backgroundColor: '#4CAF50', color: 'white' }}
+        >
+          Start Grid Session
+        </button>
+      </div>
+
+      <div>
+        <p>Session ID: {sessionId || 'No session created'}</p>
+        <p>Replica ID: {replicaId || 'No replica created'}</p>
+        <p>
+          Presigned URL:{' '}
+          {presignedUrl.substring(0, 30) +
+            (presignedUrl.length > 30
+              ? ' ... ' +
+                presignedUrl.substring(
+                  presignedUrl.length - 10,
+                  presignedUrl.length,
+                )
+              : '') || 'No URL generated'}
+        </p>
+        <p>
+          WebSocket Status:{' '}
+          <span style={{ color: isConnected ? 'green' : 'red' }}>
+            {connectionStatus}
+          </span>
+        </p>
+      </div>
+      {/* Message Log */}
+      <div
+        style={{
+          margin: '10px 0',
+          padding: '10px',
+          border: '1px solid #ccc',
+          maxHeight: '200px',
+          overflowY: 'auto',
+        }}
+      >
+        <h4>Server Message Log</h4>
+        {messages.map((message, index) => (
+          <div key={index} style={{ fontSize: '12px', marginBottom: '5px' }}>
+            {message}
+          </div>
+        ))}
+      </div>
+      {!gridReady && <h3>Loading grid...</h3>}
+      {gridReady && (
+        <div>
+          <button onClick={commit}>Commit</button>
+          <button onClick={cancel}>Cancel</button>
+          <DataSheetGrid
+            value={rowValues}
+            columns={colValues}
+            rowKey="_rowId"
+            rowClassName={({ rowData }: any) => {
+              if (deletedRowIds.has(rowData._rowId)) {
+                return 'row-deleted'
+              }
+              if (createdRowIds.has(rowData._rowId)) {
+                return 'row-created'
+              }
+              if (updatedRowIds.has(rowData._rowId)) {
+                return 'row-updated'
+              } else return ''
+            }}
+            createRow={addRowToModel}
+            duplicateRow={({ rowData }: any) => ({
+              ...rowData,
+              _rowId: genId(),
+            })}
+            onChange={(newValue: any, operations: any) => {
```

Comment:
Do these have better types than `any` provided by react-datasheet-grid?

Can we pull this function out of the JSX and define it before the `return`?